qiskit_nature/second_q/operators/polynomial_tensor.py

```diff
@@ -0,0 +1,119 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Polynomial Tensor class"""
+
+from __future__ import annotations
+from typing import Dict
+from numbers import Number
+import numpy as np
+from qiskit.quantum_info.operators.mixins import (
+    LinearMixin,
+    AdjointMixin,
+    TolerancesMixin,
+)
+
+
+class PolynomialTensor(LinearMixin, AdjointMixin, TolerancesMixin):
+    """Polynomial Tensor class"""
+
+    def __init__(self, data: Dict[str, np.ndarray]):
```

Comment:
```suggestion
    def __init__(self, data: Dict[str, complex | np.ndarray]):
```
I'd like to support constant values in this class, too.
This should look like this:
```python
PolynomialTensor({
    "": 1.0,
    "+-": np.eye(2),
})
```

Reply:
We should also add `sparse` matrix support, but we can do that in the future and not necessarily directly as part of this PR :+1:

Reply:
Do we want the constants to be inside the dictionary or have them as a different argument all together?

Reply:
For one, I would like them to be in the dictionary.
However, we could also envision them as an additional argument.

The difference would be, how they get handled later:
- constant terms inside the dict would get mapped to an all-identity operation inside the SparseLabelOp that gets build from the PolynomialTensor and, by extension, will also get mapped to an all-identity Pauli term
- additional constants as extra arguments should NOT get added to the operator down the line. They would effectively be a constant eigenvalue offset, which will not be measured and, thus, not incur any noise from the measurement

For now, let's only support constants as part of the dict. I think the extra offsets as arguments will mostly be needed by classes which contain a `PolynomialTensor`, not by the `PolynomialTensor` itself

Reply:
You want to support constants two ways here potentially. I had been thinking for it outside of the dict but still getting mapped to the identity. I think if you want constants that are not being mapped they can be handled outside of the class and stored in conjunction with some instance. Are you going to allow any key value for the constant or just an empty string? You say "them" as constants do you envision more than one. I had figured it would just be one value.

Reply:
Yes, sorry, I meant a single constant value, too

I agree, a constant which is not to be mapped, should be handled outside of `PolynomialTensor`. All I meant is, that if in the future we see need for it, I would expect that kind of constant outside the dict.

I don't see a good reason for storing the constant outside the dict, when it is supposed to get mapped to the operator. We are iterating over the dict anyways, so why would we force us to look up another attribute if we can get the constant "for free" while iterating the dict items?

Reply:
So always an empty key for this - hence it can only be one value and our signal its a constant. While it can be in the dict its kinda the odd one out somewhat. We can put it as a numpy array even though its a single value so things are more consistent perhaps - but internal checks over shapes/dims will have to special case this. In term of accessing the information, maybe it will be clearer that its more consistent as such in the dictionary once we define how we want to expose things

Reply:
Yes, the shape checking would need to handle this specifically. But I think that is worth the overhead on the internal site with the benefit of having one unified iteration over `_data` available (i.e. not having to check another `constant` attribute of the class instance after having iterated over the data)

Reply:
I guess I have been thinking random access to the data, not iterator, though we could do both! Ie you get the keys you can get any matrix you want, using any of the keys, and you can get the constant at any time. If we expose it all via an iterator I guess the consumer still has to deal with matrix versus a constant as the value - I guess they see the empty key and switch on that,

Reply:
That's right. I had not thought of random access, but that might actually be possible like this, too:
```python
p_t = PolynomialTensor({
    "": 1.0,
    "+": np.arange(-2, 0),
    "+-": np.arange(2, 6).reshape((2, 2)),
})

p_t[()]  # 1.0
p_t[(0,)]  # -2.0
p_t[(1,)]  # -1.0
p_t[(0, 0)]  # 2.0
p_t[(0, 1)]  # 3.0
# etc.
```

Reply:
Yes I had thought about it at that level since it would not expose the matrices directly - but an element at a time made me wonder... maybe externally thats all one ever needs. I had thought the key might be needed too as part of the index, but if keys cannot ever be the same length, ie same dimension matrix, then there is no ambiguity without it.

Reply:
Oh hold on, there may actually be keys of identical length but different content :/ yeh, then the above does not work